Update TXM confirmation logic to use the mined transaction count

amit-momin opened the pull request:
[BCFR-620](https://smartcontract-it.atlassian.net/browse/BCFR-620)
- Confirmer changes
  - Confirmer uses the mined transaction count to determine if transactions have been re-org'd or confirmed
  - Confirmer no longer sets transaction states to `confirmed_missing_receipt`. This state is maintained in queries for backwards compatibility

- Finalizer Changes
  - Finalizer now responsible for fetching and storing receipts for confirmed transactions
  - Finalizer now responsible for resuming pending task runs
  - Finalizer now responsible for marking old transactions without receipts broadcasted before the finalized head as fatal


[BCFR-620]: https://smartcontract-it.atlassian.net/browse/BCFR-620?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

amit-momin committed:
Updated confirmation logic to use the mined transaction count

amit-momin committed:
Added changeset

amit-momin committed:
Fixed linting

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,156 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+// CheckForConfirmation fetches the mined transaction count for each enabled address and marks transactions with a lower sequence as confirmed and ones with equal or higher sequence as unconfirmed
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CheckForConfirmation(ctx context.Context, head types.Head[BLOCK_HASH]) error {
+	for _, fromAddress := range ec.enabledAddresses {
+		// Returns the total transaction count for from address which is the highest mined sequence + 1
```
nit: Let's drop this comment. This RPC call is straightforward, and the functionality is already described at the top.

amit-momin replied:
Will do, just added it in there because `SequenceAt` is sort of a misnomer

dimriou commented on .changeset/late-windows-clean.md:
```diff
@@ -0,0 +1,5 @@
+"chainlink": minor
+---
+
+Updated the TXM confirmation logic to use the mined transaction count to identify re-org'd or confirmed transactions. #internal
```
This might need to be updated. We'll have to collect all the changes and figure out the impact. For example, we're indirectly deprecating the `ConfirmedMissingReceipt` type.

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -137,14 +189,28 @@ func (f *evmFinalizer) DeliverLatestHead(head *evmtypes.Head) bool {
 func (f *evmFinalizer) ProcessHead(ctx context.Context, head *evmtypes.Head) error {
 	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
 	defer cancel()
+	// Fetch the latest finalized block
```
nit: We don't need this

huangzhen1997 commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,156 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+		}
 
-	ec.lggr.Debugw(fmt.Sprintf("Fetching receipts for %v transaction attempts", len(attempts)), "blockNum", blockNum)
+		if etx.State == TxFinalized {
```
nit:
maybe have this check at beginning of the function as well ? next to the `len(etx.TxAttempts) == 0`

amit-momin replied:
This check is just to differentiate between the different log messages. Think we still want to keep this down here so the `logValues` has all of the fields we want to log

huangzhen1997 commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,156 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
-	}
+		// nil check on receipt interface
+		if receipt != nil {
+			logValues = append(logValues,
```
nit:
you could remove this if branch, and move appending log inside here 
```
if len(attempt.Receipts) > 0 {
	receipt = attempt.Receipts[0]
}
```

amit-momin replied:
Good call! Updated in the new commit

amit-momin committed:
Addressed feedback and fixed linting

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

huangzhen1997 commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,155 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
-				continue
-			}
+
+		// Rebroadcast the one with the highest gas price
```
just curious, is TxAttempts ordered

amit-momin replied:
We do in [LoadTxesAttempts](https://github.com/smartcontractkit/chainlink/blob/a3f97b85764f5862437247ea4d03219279588586/core/chains/evm/txmgr/evm_tx_store.go#L644). Whenever we load attempts, we order them latest to oldest using the gas price/gas tip cap in descending order

amit-momin committed:
Fixed linting

amit-momin committed:
Addressed feedback

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed VRF v2 integration tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed linting

amit-momin committed:
Removed misleading log and fixed vrf v2 integration tests

amit-momin committed:
Updated receipt check for vrf v2 integration tests

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -2109,35 +1898,138 @@ func (o *evmTxStore) UpdateTxAttemptBroadcastBeforeBlockNum(ctx context.Context,
+	query := `
+		SELECT evm.tx_attempts.* FROM evm.tx_attempts
+		JOIN evm.txes ON evm.txes.ID = evm.tx_attempts.eth_tx_id
+		WHERE evm.tx_attempts.state = 'broadcast' AND evm.txes.state IN ('confirmed', 'confirmed_missing_receipt', 'fatal_error') AND evm.txes.evm_chain_id = $1 AND evm.txes.ID NOT IN (
```
Is `confirmed_missing_receipt` relevant anymore?

amit-momin replied:
It is for now for backwards compatibility for existing transactions in this state. These changes should enable us to deprecate the `confirmed_missing_receipt` state but I didn't want to include that effort in this change.

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,151 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
-	if err := ec.txStore.MarkAllConfirmedMissingReceipt(ctx, ec.chainID); err != nil {
-		return fmt.Errorf("unable to mark txes as 'confirmed_missing_receipt': %w", err)
+	// Mark the transactions included on-chain with a purge attempt as fatal error with the terminally stuck error message
+	if err := ec.txStore.UpdateTxFatalError(ctx, purgeTxIDs, ec.stuckTxDetector.StuckTxFatalError()); err != nil {
```
I'm wondering if this functionality is better suited for the Finalizer instead

amit-momin replied:
I think we would want to mark tx as fatal for being stuck as soon as possible which is why this was left in the Confirmer. Otherwise, we would only surface the fatal error after finalization which could take a long time. I also wanted to avoid marking a fatal stuck transaction as confirmed at any point to avoid any confusion with what that state implies.

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,151 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+	var confirmedTxIDs []int64
+	for _, tx := range includedTxs {
+		isPurgeTx := false
+		for _, attempt := range tx.TxAttempts {
```
I feel this could be a transaction method like `isPurgable`.

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -280,16 +348,290 @@ func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, block
+	if receipt == nil {
+		// NOTE: This should never happen, but it seems safer to check
+		// regardless to avoid a potential panic
+		l.AssumptionViolation("got nil receipt")
```
`AssumptionViolation` seems a bit excessive here, as empty responses on certain fields can happen occasionally.

amit-momin replied:
I left this untouched from what we had previously in the Confirmer but I'm not against updating this. Is `Error` or `Warn` maybe better?

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -280,16 +348,290 @@ func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, block
-func (f *evmFinalizer) buildReceiptIdList(finalizedReceipts []Receipt) []int64 {
-	receiptIds := make([]int64, len(finalizedReceipts))
+func (f *evmFinalizer) FetchAndStoreReceipts(ctx context.Context, head, latestFinalizedHead *evmtypes.Head) error {
+	attempts, err := f.txStore.FindAttemptsRequiringReceiptFetch(ctx, f.chainID)
```
Is there a way to separate an overflowed from a fatal transaction? Although they are both classified as fatal, the former has a receipt, meaning this check will eventually stop. For the latter, seems like we'll try to fetch receipts indefinitely.

amit-momin replied:
I believe that's implicitly separated by us returning attempts. Other transactions marked as fatal in the Broadcaster delete their attempts so this query wouldn't pick those up.

amit-momin committed:
Fixed CCIP integration tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Updated tests for new head type

amit-momin committed:
Fixed linting

amit-momin committed:
Fixed VRF v2 integration tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed pipeline integration tests and added logs to nonce tracker

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed VRF e2e tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed VRF e2e test

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Updated find attempts for receipt fetch query

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
Note for future changes: we might want to consider having a single state transition method instead of having one method per state.

amit-momin replied:
I agree but it gets tricky for some state transitions that require other atomic changes like adding an error message for `fatal_error` and also removing the nonce/attempts in some cases. Or adding receipts for `finalized`

dimriou commented on common/txmgr/types/tx.go:
```diff
@@ -338,6 +338,15 @@ func (e *Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) GetChecker() (Transm
 	return t, nil
 }
 
+func (e *Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) IsPurged() bool {
```
nit: I feel `IsPurged` is a bit misleading as it implies the tx has already been purged - confirmed on-chain with an empty attempt. Some alternatives: `IsPurgeable`, `HasPurgeAttempt`, or something else.

dimriou commented on common/txmgr/types/tx_store.go:
```diff
@@ -33,12 +33,6 @@ type TxStore[
 	TxHistoryReaper[CHAIN_ID]
 	TransactionStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, SEQ, FEE]
 
-	// Find confirmed txes beyond the minConfirmations param that require callback but have not yet been signaled
-	FindTxesPendingCallback(ctx context.Context, blockNum int64, chainID CHAIN_ID) (receiptsPlus []ReceiptPlus[R], err error)
-	// Update tx to mark that its callback has been signaled
-	UpdateTxCallbackCompleted(ctx context.Context, pipelineTaskRunRid uuid.UUID, chainId CHAIN_ID) error
-	SaveFetchedReceipts(ctx context.Context, r []R, state TxState, errorMsg *string, chainID CHAIN_ID) error
-
```
Let's keep these for now so we can keep track of every tx_store method required by the current TXM version and quickly compare it with TXMv2

amit-momin replied:
Couldn't we still do that through the git history? Feel like I should clean this out with this change since they aren't being used anymore. Otherwise we'd also have to maintain the methods in `evm_tx_store.go` which seems excessive

dimriou replied:
These methods are required by the Finalizer, so they have to be maintained by the `evm_tx_store.go`. I think the debate is whether Finalizer's tx store interface is enough, or whether we should also declare them in this interface for housekeeping.

amit-momin replied:
Oh sorry I thought you meant keeping all of the tx store methods I removed. Missed that you were specifically referring to these 3. I don't mind adding these back.

dimriou commented on core/internal/features/features_test.go:
```diff
@@ -631,7 +646,7 @@ observationSource   = """
 		output := outputs[0]
 		receipt := output.(map[string]interface{})
-		assert.Equal(t, "0x11", receipt["blockNumber"])
+		assert.Equal(t, "0x13", receipt["blockNumber"])
```
Did we change testing behaviour?

amit-momin replied:
I did to account for the gap between confirmation and receipt fetch. Since we can't resume task runs in the same block that we mark a transaction as confirmed, we have to commit new heads to trigger the finalizer loop to find/store receipts and resume tasks.

dimriou commented on .changeset/late-windows-clean.md:
I wouldn't classify this as a `minor` update, as this is a significant change. We need wider visibility from RelEng.

amit-momin replied:
I think marking this as `major` change might be too extreme since this is backwards compatible. I agree we probably want better visibility on this though but think we should do that separately from the semantic versioning.

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,144 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+	for _, fromAddress := range ec.enabledAddresses {
+		minedTxCount, err := ec.client.SequenceAt(ctx, fromAddress, nil)
+		if err != nil {
+			return fmt.Errorf("unable to fetch mined transaction count for address %s: %w", fromAddress.String(), err)
```
It makes more sense to use `continue` and create a `multierr` here but since multiple enabled addresses are rarely used and there is an upcoming refactoring, we can skip for now.

amit-momin replied:
I think this was small enough of a change that was worth to address now. Even though it's rare, don't want to risk introducing any issues.

dimriou replied:
The change looks good, but I would say the complete solution would also include handling the returned error because now the `processHead` loop will exit even if a single address fails. One solution could be to return the successful addresses and pass them to `ProcessStuckTransactions` and `RebroadcastWhereNecessary`. Alternatively, we can argue that we want to run `ProcessStuckTransactions` and `RebroadcastWhereNecessary` regardless of the error returned from `CheckForConfirmation`. I think for now this looks ok, so let's not waste too much time on it.

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -2109,35 +1898,138 @@ func (o *evmTxStore) UpdateTxAttemptBroadcastBeforeBlockNum(ctx context.Context,
+
+// FindReorgOrIncludedTxs finds transactions that have either been re-org'd or included on-chain based on the mined transaction count
+// If the mined transaction count receeds, transactions could have beeen re-org'd
+// If it proceeds, transactions could have been included
```
Can you add another comment here mentioning that this check assumes transactions are broadcasted in ascending order and not out of order?

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -294,178 +260,144 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
-		likelyConfirmedCount := len(likelyConfirmed)
-		if likelyConfirmedCount > 0 {
-			likelyUnconfirmedCount := len(attempts) - likelyConfirmedCount
+		if attempt.State != txmgrtypes.TxAttemptBroadcast {
```
What do we get from this check besides detection? Would there be a problem to rebroadcast this attempt?

amit-momin replied:
I don't think there would be an issue rebroadcasting this attempt. This validation carried over from the previous logic because we used to restrict the rebroadcast attempt to have to be broadcasted in `updateEthTxAttemptUnbroadcast`. But this helped me realize, we could potentially have in-progress attempts for transactions we mark as confirmed if the TXM goes down at the right time. I should mark any in-progress tx as broadcast when I mark tx as confirmed. Then I can also remove this check too.

dimriou commented on core/chains/evm/txmgr/evm_tx_store_test.go:
```diff
@@ -1905,11 +1838,149 @@ func TestORM_UpdateTxesFinalized(t *testing.T) {
+	txStore := cltest.NewTestTxStore(t, db)
+	kst := cltest.NewKeyStore(t, db)
+	_, fromAddress := cltest.MustInsertRandomKey(t, kst.Eth())
+	t.Run("finds transactions included on-chain using the mined nonce", func(t *testing.T) {
```
I think this is copy pasted

amit-momin committed:
Addressed feedback

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Cleaned config interfaces and added back methods to txstore interface

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

dimriou commented on common/txmgr/confirmer.go:
```diff
@@ -103,10 +78,6 @@ var (
-
 // Confirmer is a broad service which performs four different tasks in sequence on every new longest chain
 // Step 1: Mark that all currently pending transaction attempts were broadcast before this block
 // Step 2: Check pending transactions for receipts
```
These need to be updated

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -1149,74 +1068,39 @@ SELECT * FROM evm.txes WHERE from_address = $1 AND nonce = $2 AND state IN ('con
 }
 
-func (o *evmTxStore) UpdateTxForRebroadcast(ctx context.Context, etx Tx, etxAttempt TxAttempt) error {
+func (o *evmTxStore) UpdateTxForRebroadcast(ctx context.Context, etxIDs []int64, attemptIDs []int64) error {
```
nit:
```suggestion
func (o *evmTxStore) UpdateTxsForRebroadcast(ctx context.Context, etxIDs []int64, attemptIDs []int64) error {
```

dimriou commented on core/chains/evm/txmgr/confirmer_test.go:
```diff
@@ -193,1249 +183,327 @@ func TestEthConfirmer_Lifecycle(t *testing.T) {
-			// Second transaction still unconfirmed
-			elems[1].Result = &evmtypes.Receipt{}
-		}).Once()
+	t.Run("handles multiple transactions at a time", func(t *testing.T) {
```
There is another test case below with the same name, but slightly different behaviour. Perhaps rename one of the two.

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -280,16 +348,290 @@ func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, block
+	}
+
+	// Mark transactions as fatal error
+	err = f.txStore.UpdateTxFatalError(ctx, oldTxIDs, ErrCouldNotGetReceipt)
```
Shouldn't we call `UpdateTxFatalErrorAndDeleteAttempts` instead?

amit-momin replied:
By calling `UpdateTxFatalError` instead, I wanted to preserve some history that this tx had attempts and never made it on-chain. Otherwise, `UpdateTxFatalErrorAndDeleteAttempts` also clears other fields like nonce which I think we want to preserve so our local state shows that nonce as being used. Neither of these might really matter but seemed better to hold on to info rather than clear it here.

dimriou replied:
I thought the only distinction between overflowed and regular fatal transactions was the fact that the first ones don't have attempts, otherwise, we would retry to fetch receipts for them.

amit-momin replied:
Oh right good point, just updated the logic to use `UpdateTxFatalErrorAndDeleteAttempts`

amit-momin committed:
Updated confirmer description and txstore method name

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Changed logic to mark old transactions without receipts as fatal

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

huangzhen1997 commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -2109,35 +1898,147 @@ func (o *evmTxStore) UpdateTxAttemptBroadcastBeforeBlockNum(ctx context.Context,
+			JOIN evm.tx_attempts ON evm.tx_attempts.eth_tx_id = evm.txes.ID
+			JOIN evm.receipts ON evm.receipts.tx_hash = evm.tx_attempts.hash
+			WHERE evm.txes.evm_chain_id = $1 AND evm.txes.state IN ('confirmed', 'confirmed_missing_receipt', 'fatal_error') AND evm.receipts.ID IS NOT NULL
+		)
```
I am curious if we can use another join to replace the `NOT IN (....)`, since the subquery seems a bit complicated

amit-momin replied:
Dimitris suggested the same earlier but I wasn't able to simplify it with another join. Don't think I can just add JOIN on the receipt table with a `evm.receipts.ID IS NOT NULL` condition. That would still include attempts for transactions that have receipts. But if we find during testing this query is too heavy, I could consider it making the change and be ok with the extra RPC calls since it's in a batch anyways.

huangzhen1997 commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -958,21 +936,13 @@ func (o *evmTxStore) SaveFetchedReceipts(ctx context.Context, r []*evmtypes.Rece
+		state = 'broadcast',
+		broadcast_before_block_num = COALESCE(evm.tx_attempts.broadcast_before_block_num, inserted_receipts.block_number)
+	FROM inserted_receipts
+	WHERE inserted_receipts.tx_hash = evm.tx_attempts.hash
```
did u miss the evm_chain_id ? otherwise you can remove `chainID` it's unused

amit-momin replied:
Oh ya good point. We don't need it for this update since tx_hash is unique but I can remove chainID from this method

amit-momin committed:
Removed chain ID from the SaveFetchedReceipts method

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

huangzhen1997 commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -2109,35 +1898,147 @@ func (o *evmTxStore) UpdateTxAttemptBroadcastBeforeBlockNum(ctx context.Context,
+	if len(txHashes) == 0 {
 		return nil
 	}
+	txHashBytea := make([][]byte, len(txHashes))
```
is `txHashBytea` a typo ?

amit-momin replied:
I think we're referring to the sql `bytea` type since this is in the tx store

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed tests and interfaces after merge conflict

amit-momin committed:
Updated finalizer logs

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed linting

amit-momin committed:
Fixed linting

amit-momin committed:
Fixed linting

amit-momin committed:
Pre-allocate slices for linting

amit-momin committed:
Cleaned up some eventually's in vrf v2 tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Added finalized transaction count prom metric

amit-momin committed:
Fixed vrf v2 integration test

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed vrf integration test

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Updated evm store mocks

amit-momin committed:
Generated new mock and clean linting

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Added no lint comments

amit-momin committed:
Fixed ifElseChain linting

amit-momin committed:
Changed filter timer in vrf smoke test

amit-momin committed:
Undid timer change and switched to hardcoded 10s filter timeout

amit-momin committed:
Extended randomness event timeout

amit-momin committed:
Reverted hardcoded timeouts

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Removed allowed logs

amit-momin committed:
Increased integration test finality depth

amit-momin committed:
Updated vrf integration test tomls and reverted the default toml

amit-momin committed:
Updated finality depth for simulated chain config defaults in integration tests

amit-momin committed:
Replaced gomega with require for two VRF test helper methods

amit-momin committed:
Added simualted chain node configs and reverted default toml changes

amit-momin committed:
Reduced vrf simulated test configs to fallback to defaults

amit-momin committed:
Updated VRF simulated test config to match smoke defaults

amit-momin committed:
Added HistoryDepth config to vrf smoke test configs

amit-momin committed:
Removed VRF test simulates chain configs

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Reverted vrf test changes

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Fixed flakey vrf bhs integration tests

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

amit-momin committed:
Updated finalization logic to delete stale receipts if detected

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -1152,74 +1072,47 @@ SELECT * FROM evm.txes WHERE from_address = $1 AND nonce = $2 AND state IN ('con
-	}
-	_, err := orm.q.ExecContext(ctx, `UPDATE evm.tx_attempts SET broadcast_before_block_num = NULL, state = 'in_progress' WHERE id = $1`, attempt.ID)
-	return pkgerrors.Wrap(err, "updateEthTxAttemptUnbroadcast failed")
+func updateEthTxAttemptUnbroadcast(ctx context.Context, orm *evmTxStore, attemptIDs []int64) error {
```
nit:
```suggestion
func updateEthTxAttemptsUnbroadcast(ctx context.Context, orm *evmTxStore, attemptIDs []int64) error {
```

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -1152,74 +1072,47 @@ SELECT * FROM evm.txes WHERE from_address = $1 AND nonce = $2 AND state IN ('con
-	}
-	_, err := orm.q.ExecContext(ctx, `UPDATE evm.txes SET state = 'unconfirmed' WHERE id = $1`, etx.ID)
-	return pkgerrors.Wrap(err, "updateEthTxUnconfirm failed")
+func updateEthTxUnconfirm(ctx context.Context, orm *evmTxStore, etxIDs []int64) error {
```
nit:
```suggestion
func updateEthTxsUnconfirm(ctx context.Context, orm *evmTxStore, etxIDs []int64) error {
```

dimriou commented on core/chains/evm/txmgr/evm_tx_store.go:
```diff
@@ -1152,74 +1072,47 @@ SELECT * FROM evm.txes WHERE from_address = $1 AND nonce = $2 AND state IN ('con
 	_, err = orm.q.ExecContext(ctx, `
 DELETE FROM evm.receipts
 USING evm.tx_attempts
 WHERE evm.receipts.tx_hash = evm.tx_attempts.hash
```
Probably out of scope but this WHERE condition seems irrelevant

amit-momin replied:
Hmmm seems like it but I'd rather keep the query as is since it's carrying over from the previous code in case I'm misreading it.

dimriou commented on core/chains/evm/txmgr/finalizer.go:
It's not in these changes, but can we update the following message:
`Received finalized block older than one already processed. This should never happen and could be an issue with RPCs."`
It is not correct as HeadTracker doesn't provide such a guarantee.

amit-momin replied:
What if I just remove the "This should never happen" part. Think this helped catch an RPC issue recently so I still want to warn for that possibility. So something like `Received finalized block older than one already processed. There could be an issue with one or more configured RPCs.`

dimriou replied:
I would drop the: `There could be an issue with one or more configured RPCs.` and let us do the interpretation instead.

amit-momin replied:
I've removed it in the latest commit

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -163,66 +234,83 @@ func (f *evmFinalizer) processFinalizedHead(ctx context.Context, latestFinalized
-	receiptIDs := f.buildReceiptIdList(finalizedReceipts)
-
-	err = f.txStore.UpdateTxStatesToFinalizedUsingReceiptIds(ctx, receiptIDs, f.chainId)
+	err = f.txStore.UpdateTxStatesToFinalizedUsingTxHashes(ctx, txHashes, f.chainID)
```
Nice

dimriou commented on core/chains/evm/txmgr/finalizer.go:
```diff
@@ -163,66 +234,83 @@ func (f *evmFinalizer) processFinalizedHead(ctx context.Context, latestFinalized
 	if err != nil {
 		return fmt.Errorf("failed to update transactions as finalized: %w", err)
 	}
-	// Update lastProcessedFinalizedBlockNum after processing has completed to allow failed processing to retry on subsequent heads
-	// Does not need to be protected with mutex lock because the Finalizer only runs in a single loop
-	f.lastProcessedFinalizedBlockNum = latestFinalizedHead.BlockNumber()
```
You removed this by accident

amit-momin committed:
Addressed feedback

amit-momin committed:
Updated log

amit-momin committed:
Merge branch 'develop' into BCI-4097-Update-the-TXM-confirmation-logic-to-use-the-mined-nonce

huangzhen1997 reviewed (commented):
Overall looks good to me, can you remind me the reason for moving minConfirmation callback logic to finalizer instead of confirmer?

amit-momin commented:
> Overall looks good to me, can you remind me the reason for moving minConfirmation callback logic to finalizer instead of confirmer?

Ya the `minConfirmation` logic to resume pending task runs are dependent on transaction receipts since we need the blocknum from them to determine when we can resume. Since we changed the confirmation logic in the Confirmer to use nonces, we moved the receipt fetching/storing to the finalizer. So it made sense to move the `minConfirmation` logic there too since we only need receipts for resuming runs and finalization now.
